# This is a namespace agnostic configuration file for the Franka FR3 robot.
# That is to say, it is not tied to any specific namespace and can be used in
# any namespace. 
# If you require namespace specific configurations, you can easily modify this
# file to include the namespace
# As an example, below you can find:
#
# /**:
#   franka_robot_state_broadcaster:
#     ros__parameters:
#       lock_try_count: 5
#       lock_sleep_interval: 5
#
# Imagine, you have a specific need to have the lock try count in namespace "NAMESPACE_X"
# to be different than the default value of 5, but leave all the other namespaces as default.
# Add the following lines to the configuration file 
# (or use a different config file, but then you have to mody the launch script to reference that file)
#
# /NAMESPACE_X:
#   franka_robot_state_broadcaster:
#     ros__parameters:
#       lock_try_count: 10
#       lock_sleep_interval: 35
# /**:
#   franka_robot_state_broadcaster:
#     ros__parameters:
#       lock_try_count: 5
#       lock_sleep_interval: 5


/**:
  controller_manager:
    ros__parameters:
      update_rate: 1000
      thread_priority: 98

      cartesian_elbow_example_controller:
        type: franka_example_controllers/CartesianElbowExampleController

      cartesian_orientation_example_controller:
        type: franka_example_controllers/CartesianOrientationExampleController

      cartesian_pose_elbow_example_controller:
        type: franka_example_controllers/CartesianElbowExampleController

      cartesian_pose_example_controller:
        type: franka_example_controllers/CartesianPoseExampleController

      cartesian_velocity_example_controller:
        type: franka_example_controllers/CartesianVelocityExampleController
      cartesian_velocity_controller:
        type: iitgn_controllers/CartesianVelocityController

      elbow_example_controller:
        type: franka_example_controllers/ElbowExampleController

      franka_robot_state_broadcaster:
        type: franka_robot_state_broadcaster/FrankaRobotStateBroadcaster
        state_interfaces:
          robot_state

      gravity_compensation_example_controller:
        type: franka_example_controllers/GravityCompensationExampleController

      gripper_example_controller:
        type: franka_example_controllers/GripperExampleController

      joint_impedance_example_controller:
        type: franka_example_controllers/JointImpedanceExampleController

      joint_impedance_with_ik_example_controller:
        type: franka_example_controllers/JointImpedanceWithIKExampleController

      joint_position_example_controller:
        type: franka_example_controllers/JointPositionExampleController

      joint_state_broadcaster:
        type: joint_state_broadcaster/JointStateBroadcaster

      joint_velocity_example_controller:
        type: franka_example_controllers/JointVelocityExampleController
      
      joint_position_controller:
        type: iitgn_controllers/JointPositionController

      joint_velocity_controller:
        type: iitgn_controllers/JointVelocityController

      model_example_controller:
        type: franka_example_controllers/ModelExampleController

      move_to_start_example_controller:
        type: franka_example_controllers/MoveToStartExampleController

      impedance_controller:
        type: iitgn_controllers/ImpedanceController

      gripper_topic_controller:
        type: iitgn_controllers/GripperTopicController
      move_to_start_controller:
        type: iitgn_controllers/MoveToStartController

/**:
  cartesian_elbow_example_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  cartesian_orientation_example_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  cartesian_pose_example_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  cartesian_velocity_example_controller:
    ros__parameters:
      arm_id: "fr3"   

/**:
  cartesian_velocity_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  elbow_example_controller:
    ros__parameters:
      arm_id: "fr3"
      joints:
        - fr3_joint1
        - fr3_joint2
        - fr3_joint3
        - fr3_joint4
        - fr3_joint5
        - fr3_joint6
        - fr3_joint7

/**:
  franka_robot_state_broadcaster:
    ros__parameters:
      lock_try_count: 5 # Number of attempts to acquire the lock in a single on_update()
      lock_sleep_interval: 5 # microseconds sleep between attempts
      lock_log_error: true #  Log error if lock acquisition fails - false == do not log  
      lock_update_success: false # return update OK, although lock fails (no update occurs)

/**:
  gravity_compensation_example_controller:
    ros__parameters:
      arm_id: "fr3"
      
/**:
  joint_impedance_example_controller:
    ros__parameters:
      arm_id: "fr3"    
      k_gains:
        - 24.0
        - 24.0
        - 24.0
        - 24.0
        - 10.0
        - 6.0
        - 2.0
      d_gains:
        - 2.0
        - 2.0
        - 2.0
        - 1.0
        - 1.0
        - 1.0
        - 0.5

/**:
  joint_impedance_with_ik_example_controller:
    ros__parameters:
      arm_id: "fr3"
      k_gains:
        - 600.0
        - 600.0
        - 600.0
        - 600.0
        - 250.0
        - 150.0
        - 50.0
      d_gains:
        - 30.0
        - 30.0
        - 30.0
        - 30.0
        - 10.0
        - 10.0
        - 5.0

/**:
  joint_position_example_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  joint_velocity_example_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  joint_velocity_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  impedance_controller:
    ros__parameters:
      arm_id: "fr3"
      command_topic: "/impedance_command"
      k_gains:
        - 24.0
        - 24.0
        - 24.0
        - 24.0
        - 10.0
        - 6.0
        - 2.0
      d_gains:
        - 2.0
        - 2.0
        - 2.0
        - 1.0
        - 1.0
        - 1.0
        - 0.5
      m_gains:
        - 1.0
        - 1.0
        - 1.0
        - 1.0
        - 1.0
        - 1.0
        - 1.0
      impedance_mode: "compliance_motion" # Options: "cartesian", "joint", "joint_velocity"
      compliance_factor: 0.5
      motion_amplitude: 0.3
      motion_frequency: 0.2
      target_position: [0.0, -0.785, 0.0, -2.356, 0.0, 1.571, 0.785]


/**:
  joint_state_broadcaster:
    ros__parameters:
      arm_id: "fr3"
      joints:
        - fr3_joint1
        - fr3_joint2
        - fr3_joint3
        - fr3_joint4
        - fr3_joint5
        - fr3_joint6
        - fr3_joint7

/**:
  gripper_topic_controller:
    ros__parameters:
      # Robot arm ID
      arm_id: "fr3"
      
      # Topic for receiving gripper commands
      gripper_command_topic: "~/gripper_command"
      
      # Gripper parameters
      gripper_open_width: 0.08    # meters (fully open)
      gripper_close_width: 0.04  # meters (target grasp width)
      gripper_speed: 0.1          # meters/second
      gripper_force: 10.0        # Newtons
      
      # Epsilon parameters for grasp tolerance
      epsilon_inner: 0.03        # meters (minimum acceptable grasp)
      epsilon_outer: 0.03        # meters (maximum acceptable grasp)




/**:
  joint_velocity_example_controller:
    ros__parameters:
      arm_id: "fr3"
 
/**:
  model_example_controller:
    ros__parameters:
      arm_id: "fr3"

/**:
  move_to_start_example_controller:
    ros__parameters:
      k_gains:
        - 600.0
        - 600.0
        - 600.0
        - 600.0
        - 250.0
        - 150.0
        - 50.0
      d_gains:
        - 30.0
        - 30.0
        - 30.0
        - 30.0
        - 10.0
        - 10.0
        - 5.0

/**:
  move_to_start_controller:
    ros__parameters:
      arm_id: "fr3"
      k_gains:
        - 600.0
        - 600.0
        - 600.0
        - 600.0
        - 250.0
        - 150.0
        - 50.0
      d_gains:
        - 30.0
        - 30.0
        - 30.0
        - 30.0
        - 10.0
        - 10.0
        - 5.0

/**:
  joint_position_controller:
    ros__parameters:
      # Robot identification
      arm_id: "fr3"
      
      # PID Gains for each joint (7 joints for FR3)
      # These gains need to be tuned for your specific robot and application
      kp_gains: [400.0, 400.0, 400.0, 400.0, 250.0, 150.0, 50.0]  # Proportional gains
      kd_gains: [30.0, 30.0, 30.0, 30.0, 20.0, 15.0, 5.0]        # Derivative gains  
      ki_gains: [10.0, 10.0, 10.0, 10.0, 5.0, 5.0, 2.0]          # Integral gains
      
      # Target position (radians) - optional, defaults to current position
      target_position: [0.0, -0.785, 0.0, -2.356, 0.0, 1.571, 0.785]
      
      # Control mode: position_control, trajectory_tracking, sinusoidal_motion, hold_position
      control_mode: "position_control"
      
      # Trajectory parameters
      trajectory_duration: 5.0  # Duration for smooth trajectories (seconds)
      
      # Sinusoidal motion parameters (for sinusoidal_motion mode)
      motion_amplitude: 0.5     # Amplitude in radians
      motion_frequency: 0.2     # Frequency in Hz
      
      # Command topics
      command_topic: "/joint_position_command"
      
